File: services/app/dockers/clojure/checker_example.clj
(ns checker_example
  (:require
    [clojure.test :refer :all]
    [clojure.data.json :as json]))

(try
  (load-file "solution_example.clj")
  (catch Exception ex
    (do
     (println (json/write-str {:status "error" :result (.getMessage ex)}))
     (System/exit 0)
     )))

(defn assert_result
  [expected result errorMessage]
  (try (assert (= expected result))
       true
       (catch java.lang.AssertionError e
         (do
           (println (json/write-str {:status "failure" :result result :arguments errorMessage}))
           false
           ))))

(defn get_test_status [status_test status_assert]
  (if status_test
    (if status_assert
      status_test
      status_assert
      )
    status_test))

(defn generate-tests
  [solution]
  (let [success
        (reduce get_test_status
          [
            (assert_result 3 (apply solution [1, 2]) [1, 2])
            (assert_result 8 (apply solution [5, 3]) [5, 3])
          ])]
    (if success
        (println (json/write-str {:status "ok" :result "__code-0__"})))))

(defn -main []
  (generate-tests solution))

File: services/app/dockers/elixir/checker_example.exs
try do
  Code.eval_file("./solution_example.exs")
  require Jason.Helpers

  ExUnit.start()
  # ExUnit.configure(capture_log: true)

  defmodule TheTest do
    use ExUnit.Case
    import Jason.Helpers

    defp assert_result(result, expected, errorMessage) do
      try do
        assert result == expected
      rescue
        _e in ExUnit.AssertionError ->
          message = json_map(status: :failure, result: errorMessage)
          IO.puts(Jason.encode!(message))
          exit(:normal)
      end
    end

    defp test_solution do
      assert_result(apply(Solution, :solution, [1, 2]), 3, [1, 2])
      assert_result(apply(Solution, :solution, [3, 5]), 8, [3, 5])
      message = json_map(status: :ok, result: "__code-0__")
      IO.puts(Jason.encode!(message))
    end

    test "solution" do
      test_solution()
    end
  end

  exit(:normal)
rescue
  e in CompileError ->
    message = Jason.Helper.json_map(status: :error, result: e.description)
    IO.puts(Jason.encode!(message))
end
